module/VuFind/src/VuFindTest/Integration/MinkTestCase.php

```diff
@@ -752,6 +753,127 @@ protected function logWarning(string $consoleMsg, string $logMsg = ''): void
+     * default)
+     *
+     * @return void
+     *
+     * @throws \RuntimeException
+     */
+    protected function validateHtml(?Element $page = null): void
+    {
+        if ((!$this->session && !$page)
+            || !($nuAddress = getenv('VUFIND_HTML_VALIDATOR'))
+        ) {
+            return;
+        }
+        $annotations = Test::parseTestMethodAnnotations(
+            static::class,
+            $this->getName(false)
+        );
+        if (($annotations['method']['skip_html_validation'][0] ?? false)
+            || ($annotations['class']['skip_html_validation'][0] ?? false)
+        ) {
+            return;
+        }
+
+        $http = new \VuFindHttp\HttpService();
+        $client = $http->createClient(
+            $nuAddress,
+            \Laminas\Http\Request::METHOD_POST
+        );
+        $client->setEncType(\Laminas\Http\Client::ENC_FORMDATA);
+        $client->setParameterPost(
+            [
+                'out' => 'json'
+            ]
+        );
+        $page = $page ?? $this->session->getPage();
+        $client->setFileUpload(
+            'test.html',
+            'file',
+            "<!DOCTYPE html>\n" . $page->getOuterHtml(),
+            'text/html'
+        );
+        $response = $client->send();
+        if (!$response->isSuccess()) {
+            throw new \RuntimeException(
+                'Could not validate HTML: '
+                . $response->getStatusCode() . ', '
+                . $response->getBody()
+            );
+        }
+        $result = json_decode($response->getBody(), true);
+        if (!empty($result['messages'])) {
+            $errors = [];
+            $info = [];
+            foreach ($result['messages'] as $message) {
+                if ('info' === $message['type']) {
+                    $info[] = $this->htmlValidationMsgToStr($message);
+                } else {
+                    $errors[] = $this->htmlValidationMsgToStr($message);
+                }
+            }
+            $logFile = getenv('VUFIND_HTML_VALIDATOR_LOG_FILE');
+            if ($info) {
+                $message = 'HTML validation messages for '
+                    . $this->session->getCurrentUrl() . ': ' . PHP_EOL . PHP_EOL
+                    . implode(PHP_EOL . PHP_EOL, $info);
+
+                if ($logFile) {
+                    file_put_contents(
+                        $logFile,
+                        date('Y-m-d H:i:s') . " $message" . PHP_EOL . PHP_EOL,
+                        FILE_APPEND
+                    );
+                } else {
+                    $this->logWarning($message);
+                }
+            }
+            if ($errors) {
+                $message = 'HTML validation errors for '
+                    . $this->session->getCurrentUrl() . ': ' . PHP_EOL . PHP_EOL
+                    . implode(PHP_EOL . PHP_EOL, $errors);
+
+                if (getenv('VUFIND_HTML_VALIDATOR_FAIL_TESTS') !== '0') {
```

Comment:
Should the exception throwing be independent of the logging/displaying of messages? I'm thinking we might want to log the error before we fail the test, if we have both logging and test failures turned on at once.

Reply:
Yes, revised. There's now a new html_validator_quiet option for suppressing output during tests as well.